streaming/receiver.c

```diff
@@ -51,47 +51,235 @@ static void rrdpush_receiver_thread_cleanup(void *ptr) {
+    debug(D_REPLICATION, "Replication on %s @ %ld, block %ld-%ld last_update=%ld", st->name, now,
+                         user->st->state->window_start, user->st->state->window_end, st->last_updated.tv_sec);
+
+    return PARSER_RC_OK;
+disable:
+    errno = 0;
+    error("Replication failed - Invalid REPBEGIN %s %s %s on host %s. Disabling it.", words[1], words[2], words[3],
+          user->host->hostname);
+    user->enabled = 0;
+    return PARSER_RC_ERROR;
+}
+
+PARSER_RC streaming_rep_dim(char **words, void *user_v, PLUGINSD_ACTION *plugins_action) {
+    PARSER_USER_OBJECT *user = user_v;
+    UNUSED(plugins_action);
+    char *id        = words[1];
+    char *time_txt  = words[2];
+    char *value_txt = words[3];
+
+    if (user->st == NULL) {
+        errno = 0;
+        error("Received RRDDIM for %s out of sequence", id);
+        goto disable;
+    }
+
+    if (user->st->state->ignore_block)
+        return PARSER_RC_OK;
+
+    if (!id || !time_txt || !value_txt)
+        goto disable;
+
+    time_t timestamp = (time_t)str2ul(time_txt);
+    storage_number value = str2ull(value_txt);
+
+    RRDDIM *rd = rrddim_find(user->st, id);
+    //time_t st_last = rrdset_last_entry_t(user->st);  UNUSED?
+    if (rd == NULL) {
+        errno = 0;
+        error("Unknown dimension \"%s\" on %s during replication - ignoring", id, user->st->name);
+        return PARSER_RC_OK;
+    }
+
+    if (user->st->rrd_memory_mode == RRD_MEMORY_MODE_DBENGINE)
+    {
+        rd->state->collect_ops.store_metric(rd, timestamp * USEC_PER_SEC, value);
+        debug(D_REPLICATION, "store " STORAGE_NUMBER_FORMAT "@%ld for %s.%s", value, timestamp, user->st->id, id);
+        rd->last_stored_value = value;
+    }
+    else
+    {
+        // We are assuming here that data received for each dimension in the window is dense. This is because any
+        // gaps from dropped samples should be filled with either SN_EMPTY_SLOT or interpolated values at storage time
+        // on the collecting node.
+        size_t offset = (size_t)(timestamp - user->st->last_updated.tv_sec) / user->st->update_every;
+        rd->values[(rd->rrdset->current_entry + offset) % rd->entries] = value;
+        debug(D_REPLICATION, "store " STORAGE_NUMBER_FORMAT "@%ld = %ld + %zu for %s.%s", value, timestamp,
+              rd->rrdset->current_entry, offset, user->st->id, id);
+    }
+    rd->collections_counter++;
+
+    return PARSER_RC_OK;
+disable:
+    error("Gap replication failed - Invalid REPDIM %s %s %s on host %s. Disabling it.", words[1], words[2], words[3],
+          user->host->hostname);
+    user->enabled = 0;
+    return PARSER_RC_ERROR;
+}
+
+PARSER_RC streaming_rep_end(char **words, void *user_v, PLUGINSD_ACTION *plugins_action) {
+    PARSER_USER_OBJECT *user = user_v;
+    UNUSED(plugins_action);
+    char *num_points_txt = words[1];
+    char *col_total_txt  = words[2];
+    char *last_total_txt = words[3];
+    if (!num_points_txt)
```

Comment:
You are not verifying the addresses for `col_total_txt` and `last_total_txt`, but you use them in the lines 251 and 252.

Reply:
Fixed.